### PR TITLE
Version bit rejection and RPC Cfund Tests

### PR DESCRIPTION
**Pull request by red010b37**
This PR introduces the concept of bit version rejection. 

It is designed to make it easier for the network to reject individual soft forks when they are bundled together in 1 release. As soft forks come within the software signaling by default only the converse was needed for people to reject soft forks they did not agree with.

A new config concept has been added called `rejectversionbit`  users can signal all the soft forks they reject by adding the following to the config

```
rejectversionbit=15
rejectversionbit=16
rejectversionbit=17
```

I built this as a new function in versionbits.cpp as that seemed the most logical place instead of mashing it into the ComputeBlockVersion, I think this will allows some expansion later down the line

RPC test:
This is cfund rpc coverage the Marcus and Gal have added

### CODE REVIEW

**Comment by red010b37** [comment]
@aguycalled once you are happy, I will update the release notes

File: src/versionbits.cpp [hunk 1]
```diff
@@ -136,6 +136,27 @@ class VersionBitsConditionChecker : public AbstractThresholdConditionChecker {
...
+    bool isRejected = false;
+
+    std::vector<std::string>& versionBitVotes = mapMultiArgs["-rejectversionbit"];
+
+    int bitTest = params.vDeployments[pos].bit;
+
+     BOOST_FOREACH(std::string rejectedBit, versionBitVotes) {
+       int rBit =  stoi(rejectedBit);
+
+       if(rBit == bitTest) {
```
Comment (proletesseract) [inline src/versionbits.cpp:150]: From what I understand reading the code we are looping through all the deployments defined in chainparams.cpp which is good. we don't need to test if the number is out of range or if the bit exists.

Are there any version bits which we don't want to be disabled in this way? Is there any ill effect if the network turns off any of the other version bits?

https://github.com/aguycalled/navcoin-core/blob/bc765f242c2a59f684068eeef323fff376f34e5a/src/chainparams.cpp#L126

I assume it will have no effect, but wouldn't mind a second opinion @aguycalled

File: src/versionbits.cpp [hunk 2]
```diff
@@ -136,6 +136,27 @@ class VersionBitsConditionChecker : public AbstractThresholdConditionChecker {
...
+bool IsVersionBitRejected(const Consensus::Params& params, Consensus::DeploymentPos pos){
+
+    bool isRejected = false;
+
+    std::vector<std::string>& versionBitVotes = mapMultiArgs["-rejectversionbit"];
+
+    int bitTest = params.vDeployments[pos].bit;
+
+     BOOST_FOREACH(std::string rejectedBit, versionBitVotes) {
+       int rBit =  stoi(rejectedBit);
```
Comment (proletesseract) [inline src/versionbits.cpp:148]: what happens if i don't have an int string in my config eg. rejectversionbit=A

Do we care?

File: src/main.cpp [hunk 3]
```diff
@@ -2472,7 +2472,7 @@ int32_t ComputeBlockVersion(const CBlockIndex* pindexPrev, const Consensus::Para
 
     for (int i = 0; i < (int)Consensus::MAX_VERSION_BITS_DEPLOYMENTS; i++) {
         ThresholdState state = VersionBitsState(pindexPrev, params, (Consensus::DeploymentPos)i, versionbitscache);
-        if (state == THRESHOLD_LOCKED_IN || state == THRESHOLD_STARTED || state == THRESHOLD_ACTIVE) {
+        if (state == THRESHOLD_LOCKED_IN || state == THRESHOLD_STARTED || state == THRESHOLD_ACTIVE || !IsVersionBitRejected(params, (Consensus::DeploymentPos)i)) {
```
Comment (proletesseract) [inline src/main.cpp:2475]: should this be && instead of an || ? Otherwise just one of these things need to return true and the version bit will be concatenated. 
```
if ((state == THRESHOLD_LOCKED_IN || state == THRESHOLD_STARTED || state == THRESHOLD_ACTIVE) && !IsVersionBitRejected(params, (Consensus::DeploymentPos)i)) {
...
}
```
Or am i misunderstanding this?

File: src/versionbits.cpp (hunk 1 above)
Comment (red010b37) [reply src/versionbits.cpp:150]: Im not to sure @aguycalled ?

File: src/versionbits.cpp (hunk 2 above)
Comment (red010b37) [reply src/versionbits.cpp:148]: im not to sure i can try, catch it or something? again @aguycalled

**Comment by aguycalled** [comment]
ideally the new cfund rpc tests should be in a separate pr/direct commit to the cfund branch

File: src/versionbits.cpp (hunk 2 above)
Comment (aguycalled) [reply src/versionbits.cpp:148]: it would be interpreted as 0. isdigit() can be used as a very simple check, other than that can be unnecessarily resource-consuming
http://www.cplusplus.com/reference/cctype/isdigit/

File: src/main.cpp (hunk 3 above)
Comment (aguycalled) [reply src/main.cpp:2475]: agree with @craigmacgregor

File: src/versionbits.cpp (hunk 1 above)
Comment (aguycalled) [reply src/versionbits.cpp:150]: from what i read in the code, he's looping through the different values given to the key "rejectversionbit" in the config file. the only optimisation i can see here would be to directly return true whenever a match is found, to save from going through the next iterations. the necessary bit masks are set independent from this check, so i don't consider any risk

**Comment by red010b37** [comment]
I have created a new PR with the requested changes and only the version bit code

New PR: https://github.com/aguycalled/navcoin-core/pull/15
